lib/spack/llnl/util/argparsewriter.py

```diff
@@ -10,195 +10,345 @@
 import sys
 
 
+# NOTE: The only reason we subclass argparse.HelpFormatter is to get access
+# to self._expand_help(), ArgparseWriter is not intended to be used as a
+# formatter_class.
 class ArgparseWriter(argparse.HelpFormatter):
     """Analyzes an argparse ArgumentParser for easy generation of help."""
-    def __init__(self, out=sys.stdout):
-        super(ArgparseWriter, self).__init__(out)
+
+    def __init__(self, prog, out=sys.stdout):
+        """Initializes a new ArgparseWriter instance.
+
+        Parameters:
+            prog (str): the program name
+            out (file object): the file to write to
+        """
+        super(ArgparseWriter, self).__init__(prog)
         self.level = 0
         self.out = out
 
-    def _write(self, parser, root=True, level=0):
+    def parse(self, parser):
+        """Parses the parser object and returns the relavent components.
+
+        Parameters:
+            parser (argparse.ArgumentParser): the parser
+
+        Returns:
+            str: the command name
+            str: the command description
+            str: the command usage
+            list: list of positional arguments
+            list: list of optional arguments
+            list: list of subcommand parsers
+        """
         self.parser = parser
-        self.level = level
-        actions = parser._actions
 
-        # allow root level to be flattened with rest of commands
-        if type(root) == int:
```

Comment:
I got rid of the `root` parameter to `write`. The only consequence of this is that `spack` will be one level higher in the [Command Reference](https://spack.readthedocs.io/en/latest/command_index.html#spack). If this is undesirable, we can either try to add it back, or just replace it after the fact/hardcode it.

Reply:
We could actually replace `=` with `-` in `_rst_levels` if we want. I think that's a better option, let me do that.